### ISSUE TITLE
VS/DR can share same host in some cases

### ISSUE THREAD

**Issue by lucasponce** [kiali/kiali#1471]
Currently, we warn if the same host is shared by multiple Destination Rules or multiple Virtual Services. There are good reasons to it, but sadly there are also cases where this is acceptable.

Istio documentation gives an instruction to split the definition in certain cases as long as one avoids the potential pitfalls.

We should probably modify our validation processes to accept these multiple hosts in those cases where these pitfalls are not an issue. Otherwise we might see people complain that the istio documentation says "do it" and we say "don't do it".

https://istio.io/help/ops/traffic-management/deploy-guidelines/#multiple-virtual-services-and-destination-rules-for-the-same-host

JIRA:
https://issues.jboss.org/browse/KIALI-2490

**Comment by bryonwinger** [comment]
We have a use case where we have a multi-tenant application where a particular service will be used across many domains. The mesh doesn't otherwise need to be aware of the tenant and all incoming requests are sanitized outside of the mesh. This works as expected when using `*` for hosts in the Virtual Service spec, but we would really like to avoid the warnings in Kiali.

Ideally we want to avoid the need to update the VS every time we add a new host.Is there any way to configure Kiali to disable this warning (either entirely or per resource)?

**Comment by lucasponce** [comment]
Perhaps we can introduce some flag to consider wildcars as this case on Kiali validations.

**Comment by lucasponce** [comment]
@bryonwinger can you share a little bit more your details ?

I mean, I understand that you may have VirtualService A and VirtualService B with host wildcards "*".
So, I understand that you have a different gateway per VS ? or how do you resolve the potential ambiguity ?

I mean if I invoke host.test.com, both VS A and VS B will process it (unless they have different gateways).

I'm thinking loud on this to see how we can refine the validation (perhaps same wildcard but different gateways is valid, but same wildcard with same or no gateway should remain a warning).

I mean, disable a potential validation is an option, but I know what this may fall: just always keep something disabled. I don't discard this at the moment but with more info about use cases perhaps we can adapt the validation to be more useful for your use cases.

(Also, I'd like to code an example like you use, to help us testing more general scenarios).

Thanks

**Comment by lucasponce** [comment]
I've commented with @xeviknal about to relax the warning conditions in Kiali 
https://github.com/kiali/kiali/issues/3074

But I think that probably there can be cases where 2 VS sharing same host could lead in a potential problem, so adding a warning it's something that at least should inform to the user.

**Comment by bryonwinger** [comment]
@lucasponce the VirtualServices do use different gateways. Here is a simplified example:

```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: admin-gateway
spec:
  gateways:
  - admin-gateway
  hosts:
  - '*'
  http:
  - route:
    - destination:
        host: admin-gateway
        subset: a
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: admin-gateway
spec:
  host: admin-gateway
  subsets:
  - labels:
      version: a
    name: stable
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: user-gateway
spec:
  gateways:
  - user-gateway
  hosts:
  - '*'
  http:
  - route:
    - destination:
        host: user-gateway
        subset: a
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: user-gateway
spec:
  host: user-gateway
  subsets:
  - labels:
      version: a
    name: stable
```

We have many non-gateway VS too, though these are easier to be set the host for as they are just internal services. As soon as we use `*` once though, we get warnings for all of the VS on many different pages in Kiali. Since we are expecting and ok with this particular warning, it makes it hard for us to notice any other warnings that we might be concerned with.

**Comment by lucasponce** [comment]
Thanks @bryonwinger. So, just to confirm:
- '*' wildcard is used with different gateways (I wonder if perhaps we can validate the pair wildcard/gateway).
- '*' can be used as well in internal services ?

**Comment by bryonwinger** [comment]
@lucasponce we only use `*` for `spec.hosts` on a VirtualService that has a Gateway.

For other internal services, the VS just uses the name of the service for `spec.hosts` and doesn't define `spec.gateways` at all. However, even with just one of the 'gateway-based' Virtual Services using `*`, we see the warning set on all other VS objects.

Hopefully that makes sense.

**Comment by xeviknal** [comment]
IMO it makes sense to only show the warning if the VS share both the same host and the same gateway (or no gateway). In case you have two gateways listening to the same hosts/ports, there is [another validation](https://kiali.io/documentation/latest/validations/#_kia0301_more_than_one_gateway_for_the_same_host_port_combination) in place that will shout.

**Comment by lucasponce** [comment]
@xeviknal @bryonwinger yeah, I think that Kiali could easily extend the validation to the pair (hosts - wildcard included - plus gateway), then same host but different gateway should be a valid scenario.

**Comment by xeviknal** [comment]
@bryonwinger the [KIALI1106](https://kiali.io/documentation/latest/validations/#_kia1106_more_than_one_virtual_service_for_same_host) is not shown when the VS has a gateway. In the configuration you shared, I can't see this validation raising. 

I also have the [following configuration](https://github.com/xeviknal/istio-lab/blob/master/1471/issue.yaml) and the validation is not present either. Both VirtualService has `*` hosts however the Gateways has specific hosts. When one of the gateways uses the `*` host, then the [KIALI0301](https://kiali.io/documentation/latest/validations/#_kia0301_more_than_one_gateway_for_the_same_host_port_combination) arises.

I'd like to find the root cause of it and try to make things easier for you guys. Do you mind sharing your Gateways too? And just to make sure, which version of Kiali are you guys using? Which validation code would you like to disable?

**Comment by bryonwinger** [comment]
@xeviknal our Gateways use `*` hosts as well. They differ from each other based on the port:

```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: admin-gateway
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      name: https-admin-gateway
      number: 32301
      protocol: HTTPS
---
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: user-gateway
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      name: https-user-gateway
      number: 32302
      protocol: HTTPS
```

We are using Kiali v1.18.2 with Istio 1.6.5 and get `KIA1106` on all of the VirtualServices, including those without a corresponding Gateway. There are no warnings on the Gateways.

In case it helps, our non-gateway VS look like this:
```
kind: VirtualService
apiVersion: networking.istio.io/v1alpha3
metadata:
  name: some-service
spec:
  hosts:
    - some-service
  http:
    - route:
        - destination:
            host: some-service
```

Traffic goes through an AWS ALB and routing is handled primarily with listener rules that forward to the different Gateway ports.

**Comment by xeviknal** [comment]
@bryonwinger it makes sense now for me. I've prepared a PR to cover your scenario: https://github.com/kiali/kiali/pull/3102
Now, every VS is only compared with other VS using the same gateway. Therefore, the validation will be present when you have two VS(1,2) using the same host and the same gateway. It also works in case V1 uses `*` hosts and other V2, under the same gateway, has other host.
Hope it does makes sense :)

**Comment by xeviknal** [comment]
@bryonwinger the solution is in place now. It will be available in kiali v1.23 (expected to be in 3 weeks) unless you want to use the latest version (I guess available tomorrow) that it should also be pretty stable.
Thank you really really much for your feedback and helping us :+1:
